README.md
mediaburstSMS
=============

This class lets you interact with the mediaburst SMS API without 
the hassle of having to create any XML or make HTTP calls.

Installation
------------
Simply download mediaburstSMS.class.php and include it in your project

Prerequisites
-------------
PHP 5 or higher

Usage
-----

Include the class in to your project

	include('mediaburstSMS.class.php');

### Sending a message

You can send a message to a single number or an array of numbers

	// Create the SMS object
	$sms = new mediaburstSMS('username', 'password');

	// Send to a single number
	$result = $sms->Send('447123456789', 'Hello World');

	// Send to multiple numbers
	$result = $sms->Send(array('mobile_number_1', 'mobile_number_2'), 'Hello World');
	
The Send call will return an array of responses, one for each number sent to. 

For a successfull send the response will contain the following:

	Array ( 
		'to'         => '447123456789',
		'id'         => 'message_id',
		'success'    => true
	); 

Whereas if a particular number failed it will look like this:

	Array (
		'to'         => 'not_a_number',
		'error_no'   => 10,
		'error_desc' => "Invalid 'To' Parameter",
		'success'    => false 
	);

The success parameter allows you to quickly see if the message worked.

### Checking your credit

Check how many SMS credits you currently have available

	$sms = new mediaburstSMS('username', 'password');
	$balance = $sms->CheckCredit();

### Handling Errors
The mediaburstSMS class will throw exceptions if the entire call failed. 
These will be type mediaburstException for API errors and Exception 
for other errors such as unable to connect to the server.  This allows
you to handle the errors in your application.

For example sending the wrong username:

	try {
		$sms = new mediaburstSMS('wrong_username', 'password');
	} catch (mediaburstException $e) {
		echo 'SMS Exception: '.$e->getMessage();
	} catch (Exception $e) {
		echo 'Exception: '.$e->getMessage();
	}

will produce the response

	SMS Exception: Invalid Username Or Password

whereas, trying to send without an internet connection would produce

        Exception: HTTP Error calling SMS API - HTTP Status: 0 - cURL Error: Couldn't resolve host 'sms.message-platform.com'	

Advanced Usage
--------------

This class has a few additional features that some users may find useful

### Optional Parameters

The SMS library supports optional parameters for the following:

*   from [string]

    The from address displayed on a phone when they receive a message

*   long [boolean, default: true]  

    Enable long SMS. A standard text can contain 160 characters, a long SMS supports up to 459.

*   ssl [boolean, default: true]

    Use SSL when making an HTTP request to the mediaburst API

* 	proxy_host [string]

	Hostname of your proxy server

*	proxy_port [integer]

	Port number on your proxy server

### Setting Options
Options can be passed as an array when creating the sms object

	$sms = new mediaburstSMS('username', 'password', array('from'=> 'from') );

Or set individually using properties

	$sms = new mediaburstSMS('username', 'password');
	$sms->from = 'php_code';

### Global Defaults
You can define constants used by the library to save having to pass them each time it is called

* MEDIABURST_USER - API Username
* MEDIABURST_PASS - API Password

For exmaple: 

	define('MEDIABURST_USER', 'username');
	define('MEDIABURST_PASS', 'password');
	$sms = new mediaburstSMS();
	$result = $sms->Send('mobile_number', 'Hello World');

### Using a proxy server

The wrapper supports requests through an HTTP proxy server, simply set the proxy_host and 
proxy_port optional parameters.

For example:

	$sms = new mediaburstSMS('username', 'password');
	$sms->proxy_host = "proxy.example.com";
	$sms->proxy_port = 3128;


License
-------

The mediaburstSMS library is distributed under the ISC license. 
A copy of the license is included in license.txt

Contributing
------------

If you have any feedback on this class drop us an email at hello@mediaburst.co.uk

To contribute improvements, create a fork and then submit a push request on GitHub



CHANGELOG.md
# php-mediaburst-sms Changelog

## 1.4.1 (20th March 2012)
*	Escaping bad characters in the message to stop a DOMDocument error being thrown when sending an &, <, >, " or '.

## 1.4 (15th Feb 2012)
*	Add request logging support, set the log option and the XML will be output to the PHP error log
*	Allow a different class to be used for HTTP transport.	Primarily used for developing 
	plugins in systems that have their own HTTP wrappers such as WordPress

## 1.3.1 (24th Oct 2011)
*	Change internal API Urls to use api.mediaburst.co.uk 	

## 1.3 (28th Sep 2011)
*	Add support for Proxy servers, set the proxy_host and proxy_port parameters.

## 1.2 (24th Aug 2011)
*	Update CheckCredit function to use the mediaburst XML API, no change to 
	functionality, brings this wrapper in to line with API documentation.

## 1.1 (15th Aug 2011)
*   New mediaburstHTTP class
    Compatability wrapper around cURL and PHP stream interfaces for HTTP requests,
    this should make the mediaburst class work on more PHP installs without tweaking

*   Only enable SSL by default on installs where PHP supports SSL

*   Update readme to demonstrate optional parameters and clarify error handling

## 1.0 (21st Jul 2011)
* Initial public release

## 0.1 
* Internal release for http://snappr.eu


class-wordpress-mb-http.php
<?php 

/* 
 * WordPress MB HTTP class
 * 
 * Extends the mediaburstHTTP wrapper class to use the 
 * WordPress HTTP API for HTTP calls, attempts to work 
 * round the differences in PHP versions, such as SSL 
 * & curl support
 *
 * If you prefer to integrate with an existing set of HTTP 
 * functionality in your framework or CMS, you can extend
 * or completely replace this class, then pass the new
 * class name into your mediaburstSMS instance as the
 * http_class option (string).
 * 
 * @package Mediaburst SMS for WooCommerce
 * @since 1.0
 */
class WordPressMBHTTP extends mediaburstHTTP { 

	/*
	 * Check if the WordPress HTTP API can support SSL
	 *
	 * @returns bool True if SSL is supported
	 */
	public function SSLSupport() {
		return wp_http_supports( array( 'ssl' ) );
	}

	/* 
	 * Make an HTTP POST using the WordPress HTTP API.
	 *
	 * @param string url URL to send to
	 * @param string type MIME Type of data
	 * @param string data Data to POST
	 * @return string Response returned by server
	 */
	public function Post($url, $type, $data) {
		$args = array(
			'body' => $data,
			'headers' => array( 'Content-Type' => 'text/xml' ),
			'timeout' => 10, // Seconds
		);
		$result = wp_remote_post( $url, $args );
		if ( is_wp_error( $result ) ) {
			error_log( "POST failed: " . $result->get_error_message() );
			return false;
		}
		return $result[ 'body' ];
	}
}


mediaburstSMS.class.php
<?php 
/*
 * mediaburst SMS API wrapper for PHP
 * 
 * Contact hello@mediaburst.co.uk if you have any questions
 *
 * @package     mediaburstSMS
 * @author      mediaburst <hello@mediaburst.co.uk>
 * @copyright   2011 Mediaburst Ltd
 * @license     ISC
 * @version		1.4
 * @since       1.0
 * @link	http://www.mediaburst.co.uk/api/	Mediaburst API Documentation
 * @link	https://github.com/mediaburst/		Latest version of this class
 */

/* 
 * mediaburstSMS class 
 *
 * All the functionality you need to use the mediaburst SMS API is 
 * contained within this class.
 * 
 * If you use this class in multiple places throughout your project
 * you may find it easier to define your API username and password 
 * in two constants MEDIABURST_USER and MEDIABURST_PASS within an
 * include file, this saves having to pass them each time you create
 * the mediaburstSMS object
 * 
 * The latest version of this class can always be found at:
 * https://github.com/mediaburst/
 * 
 * If you have any improvements drop us an email or fork the project 
 * on GitHub and submit a push request.
 * 
 * @package	mediaburstSMS * 
 */
class mediaburstSMS {
	private $url_send = 'api.mediaburst.co.uk/xml/send.aspx';
	private $url_credit = 'api.mediaburst.co.uk/xml/credit.aspx';

	private $username;
	private $password;
	private $from;
	private $long;
	private $truncate;
	private $ssl;
	// Proxy server fields
	private $proxy_host;
	private $proxy_port;
	
	/**
	 * Enables various logging of messages when true.
	 *
	 * @var bool
	 **/
	protected $log;
	
	/**
	 * The name of the class to use for HTTP transport
	 *
	 * @var string
	 **/
	protected $http_class;
	
	/*
	 * Create a mediaburstSMS object
	 *
	 * Username and password can be passed here or defined as constants
	 * If you use the MEDIABURST_USER and MEDIABURST_PASS constants 
	 * simpy miss out these parameters when creating the object
	 * e.g. $sms = new mediaburstSMS();
	 * 
	 * @param string username	mediaburst SMS API username
	 *				Leave blank to use the MEDIABURST_USER constant 
	 * @param string password	mediaburst SMS API password
	 *				Leave blank to use the MEDIABURST_PASS constant
	 * @param array	 options	Optional parameters for sending SMS
	 */
	public function __construct( $username="", $password="", $options = array() ) {
		if( empty($username) && !defined('MEDIABURST_USER') ) 
			throw new mediaburstException("Username is blank");
		else
			$this->username = (empty($username)) ? MEDIABURST_USER : $username;
		if( empty($password) && !defined('MEDIABURST_PASS') )
			throw new mediaburstException("Password is blank"); 
		else
			$this->password = (empty($password)) ? MEDIABURST_PASS : $password;
		

		// options and defaults
		$this->long = (array_key_exists('long', $options)) ? $options['long'] : true;
		$this->from = (array_key_exists('from', $options)) ? $options['from'] : null;
		$this->truncate = (array_key_exists('truncate', $options)) ? $options['truncate'] : false;
		$this->proxy_host = (array_key_exists('proxy_host', $options)) ? $options['proxy_host'] : null;
		$this->proxy_port = (array_key_exists('proxy_port', $options)) ? $options['proxy_port'] : null;
		$this->log = (array_key_exists('log', $options)) ? $options['log'] : false;
		$this->http_class = (array_key_exists('http_class', $options)) ? $options['http_class'] : 'mediaburstHTTP';
		$this->ssl = (array_key_exists('ssl', $options)) ? $options['ssl'] : true;
	}

	/* 
	 * Sends a text message.
	 * 
	 * The response is returned as an array of responses, each 
	 * response looking similar to:
	 * Array (
	 * 		[to] => 447971687295
	 * 		[id] => VE_127890871
	 * 		[success] => 1
	 * )
	 * 
	 * @param mixed	$to	Either a string containing a single mobile 
	 *				number or an array of numbers
	 * @param string $message The text message to send
	 * @param array $add_params An associative array of additional parameters to send with the request
	 * return An array of responses, each response as an array 
	 */
	public function Send( $to, $message, $add_params = array() ) {
		// Make single number in to array for easy processing
		if (!is_array($to)) 
			$to = array($to);
    
		// Escape dodgy characters as DOMDocument throws a paddy
		$message = htmlentities( $message, ENT_QUOTES | ENT_XML1, 'UTF-8' );
		
		$req_doc = new DOMDocument('1.0', 'UTF-8');
		$root = $req_doc->createElement('Message');
		$req_doc->appendChild($root);
		$root->appendChild($req_doc->createElement('Username', $this->username));
		$root->appendChild($req_doc->createElement('Password', $this->password));

		foreach($to as $number) {
			$sms_node = $req_doc->createElement('SMS');
			$sms_node->appendChild($req_doc->createElement('To', $number)); 
			$sms_node->appendChild($req_doc->createElement('Content', $message));
			if($this->long)
				$sms_node->appendChild($req_doc->createElement('Concat', 3));
			if($this->from)
				$sms_node->appendChild($req_doc->createElement('From', $this->from));
			if($this->truncate)
				$sms_node->appendChild($req_doc->createElement('Truncate', 1));
			foreach ( $add_params as $param => $value )
				$sms_node->appendChild($req_doc->createElement($param, $value));
			$root->appendChild($sms_node);
		}

		$req_xml = $req_doc->saveXML();
		if ( $this->log )
			$this->LogXML( 'Send SMS Request', $req_xml );
		$resp_xml = $this->PostToAPI($this->url_send, $req_xml);
		if ( $this->log )
			$this->LogXML( 'Send SMS Reply', $resp_xml );
		$resp_doc = new DOMDocument();
		$resp_doc->loadXML($resp_xml);

		$resp = array(); 
		$err_no = null;
		$err_desc = null;
		foreach($resp_doc->documentElement->childNodes AS $doc_child) {
			switch($doc_child->nodeName) {
				case "SMS_Resp":
					$sms = array();
					foreach($doc_child->childNodes AS $resp_node) {
						switch($resp_node->nodeName) {
							case "To":
								$sms['to'] = $resp_node->nodeValue;
								break;
							case "MessageID":
								$sms['id'] = $resp_node->nodeValue;
								break;
							case "ErrNo":
								$sms['error_no'] = $resp_node->nodeValue;
								break;
							case "ErrDesc":
								$sms['error_desc'] = $resp_node->nodeValue;
								break;
							default:
								$sms[strtolower($resp_node->nodeName)] = $resp_node->nodeValue;
								break;
						}
					}
					$sms['success'] = !array_key_exists('error_no', $sms);
					array_push($resp, $sms);
					break;
				case "ErrNo":
					$err_no = $doc_child->nodeValue;
					break;
				case "ErrDesc":
					$err_desc = $doc_child->nodeValue;
					break;
			}
		}
		if(isset($err_no)) {
			throw new mediaburstException($err_desc, $err_no);
		}

		return $resp;
	}

	/*
	 * Check how much credit you have available on your mediaburst API account
	 *
	 * @returns	long		Number of SMS you can send
	 */
	public function CheckCredit() {
		$req_doc = new DOMDocument('1.0', 'UTF-8');
		$root = $req_doc->createElement('Credit');
		$req_doc->appendChild($root);
		$root->appendChild($req_doc->createElement('Username', $this->username));
		$root->appendChild($req_doc->createElement('Password', $this->password));
		
		$req_xml = $req_doc->saveXML();		
		if ( $this->log )
			$this->LogXML( 'Credit Balance Request', $req_xml );
		$resp_xml = $this->PostToAPI($this->url_credit, $req_xml);
		if ( $this->log )
			$this->LogXML( 'Credit Balance Reply', $resp_xml );
		
		$resp_doc = new DOMDocument();
		$resp_doc->loadXML($resp_xml);
		
		$credit;
		$err_no = null;
		$err_desc = null;
		foreach($resp_doc->documentElement->childNodes AS $doc_child) {
			switch($doc_child->nodeName) {
				case "Credit":
					$credit = $doc_child->nodeValue;
					break;
				case "ErrNo":
					$err_no = $doc_child->nodeValue;
					break;
				case "ErrDesc":
					$err_desc = $doc_child->nodeValue;
					break;
			}
		}
		if(isset($err_no)) 
			throw new mediaburstException($err_desc, $err_no);
		
		return $credit;
	}

	/*
	 * Make an HTTP POST to the mediaburst API server
	 * 
	 * @param	string	url	URL to send to
	 * @param	string	data	Data to post
	 * @return	string		Server response
	 */
	private function PostToAPI($url, $data) {
		$http_class = $this->http_class;
		$http = new $http_class();

		if($this->ssl && $http->SSLSupport())
			$url = 'https://'.$url;
		else
			$url = 'http://'.$url;
		
		$http->proxy_host = isset($this->proxy_host) ? $this->proxy_host : null;
		$http->proxy_port = isset($this->proxy_port) ? $this->proxy_port : null;

		return $http->Post($url, 'text/xml', $data);
	}

	/**
	 * Log some XML, tidily if possible, in the PHP error log
	 *
	 * @param string $log_msg The log message to prepend to the XML
	 * @param string $xml An XML formatted string
	 * @return void
	 **/
	protected function LogXML( $log_msg, $xml ) {
		// Tidy if possible
		if ( class_exists( 'tidy' ) ) {
			$tidy = new tidy;
			$config = array(
				'indent' => true,
				'input-xml'	=> true,
				'output-xml' => true,
				'wrap' => 200
			);
			$tidy->parseString( $xml, $config, 'utf8' );
			$tidy->cleanRepair();
			$xml = $tidy;
		}
		// Output
		error_log( "MBSMS $log_msg:  $xml" );
	}

	// Use PHP magic function to create GET property accessor 
	// i.e. $sms->from will call get_from() on the object
	public function __get($name) {
		if(method_exists($this, "get_$name"))
			return $this->{"get_$name"}();
		throw new Exception("get_$name does not exist");
	}
	// Use PHP magic function to create SET property accessor
	// i.e. $sms->from = $val will call set_from($val) on the object
	public function __set($name, $value) {
		if(method_exists($this, "set_$name"))
			return $this->{"set_$name"}($value);
		throw new Exception("set_$name does not exist");
	}

	/*
	 *  Property accessors for internal variables that need to be public
	 */
	private function get_from() {
		return $this->from;
	}
	private function set_from($value) {
		$this->from = $value;
	}

	private function get_long() {
		return $this->long;
	}
	private function set_long($value) {
		$this->long = $value;
	}
	
	private function get_truncate() {
		return $this->truncate;
	}
	private function set_truncate($value) {
		$this->truncate = $value;
	}

	private function get_ssl() {
		$http_class = $this->http_class;
		$http = new $http_class();
		return $this->ssl && $http->SSLSupport();
	}
	private function set_ssl($value) {
		$this->ssl = $value;
	}

	private function get_proxy_host() {
		return $this->proxy_host;
	}
	private function set_proxy_host($value) {
		$this->proxy_host = $value;
	}

	private function get_proxy_port() {
		return $this->proxy_port;
	}
	private function set_proxy_port($value) {
		$this->proxy_port = $value;
	}
}

/*
 * mediaburstException class
 *
 * The mediaburstSMS class will throw these if an error occurs
 *
 * @package     mediaburstSMS
 * @since	1.0
 */
class mediaburstException extends Exception {

        public function __construct( $message, $code=0 ) {
                // make sure everything is assigned properly
                parent::__construct($message, $code);
        }
}

/* 
 * mediaburstHTTP class
 * 
 * Wrapper class for HTTP calls, attempts to work round the 
 * differences in PHP versions, such as SSL & curl support
 * 
 * If you prefer to integrate with an existing set of HTTP 
 * functionality in your framework or CMS, you can extend
 * or completely replace this class, then pass the new
 * class name into your mediaburstSMS instance as the
 * http_class option (string).
 * 
 * @package	mediaburstSMS
 * @since	1.1
 */
class mediaburstHTTP { 
	// Optional parameters for proxy servers
	public $proxy_host;
	public $proxy_port;

	/*
	 * Check if PHP has SSL support compiled in
	 *
	 * @returns     True if SSL is supported
	 */
	public function SSLSupport() {
		$ssl = false;
		// See if PHP compiled with cURL
		if(extension_loaded('curl')) {
			$version = curl_version();
			$ssl = ($version['features'] & CURL_VERSION_SSL) ? true : false;
		} elseif (extension_loaded('openssl')) { 
			$ssl = true;
		}
		return $ssl;
	}

	/* 
	 * Make an HTTP POST 
	 * 
	 * cURL will be used if available, otherwise tries the PHP stream functions
	 * The PHP stream functions require at least PHP 5.0, cURL should work with PHP 4
	 *
	 * @param	string	url	URL to send to
	 * @param	string	type	MIME Type of data
	 * @param	string	data	Data to POST
	 * @return	string		Response returned by server
	 */
	public function Post($url, $type, $data) {
		if(extension_loaded('curl')) {
			$ch = curl_init($url);
			curl_setopt($ch, CURLOPT_POST, 1);
			curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
			curl_setopt($ch, CURLOPT_HTTPHEADER, Array("Content-Type: $type"));
			curl_setopt($ch, CURLOPT_USERAGENT, "mediaburst PHP Wrapper/1.1.0");
			curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
			if(isset($this->proxy_host) && isset($this->proxy_port)) {
				curl_setopt($ch, CURLOPT_PROXY, $this->proxy_host);
				curl_setopt($ch, CURLOPT_PROXYPORT, $this->proxy_port);
			}

			$response = curl_exec($ch);
			$info = curl_getinfo($ch);

			if ($response === false || $info['http_code'] != 200) {
				throw new Exception('HTTP Error calling SMS API - HTTP Status: '.$info['http_code'].' - cURL Erorr: '.curl_error($ch));
			} elseif(curl_errno($ch) > 0) {
				throw new Exception('HTTP Error calling SMS API - cURL Error: '.curl_error($ch));
			}

			curl_close($ch);

			return $response;
		} elseif(function_exists('stream_get_contents')) {

			// Enable error Track Errors
			$track = ini_get('track_errors');
			ini_set('track_errors',true);

			$params = array('http' => array(
				'method' => 'POST',
				'header' => "Content-Type: $type\r\nUser-Agent: mediaburst PHP Wrapper/1.1.0\r\n",
				'content' => $data
			));

			if(isset($this->proxy_host) && isset($this->proxy_port)) {
				$params['http']['proxy'] = 'tcp://'.$this->proxy_host.':'.$this->proxy_port;
				$params['http']['request_fulluri'] = True;
			}
	
			$ctx = stream_context_create($params);
			$fp = @fopen($url, 'rb', false, $ctx);
			if (!$fp) {
				ini_set('track_errors',$track);
				throw new Exception("HTTP Error calling SMS API - fopen Error: $php_errormsg");
			}
			$response = @stream_get_contents($fp);
			if ($response === false) {
				ini_set('track_errors',$track);
				throw new Exception("HTTP Error calling SMS API - Stream Error: $php_errormsg");
			}
			ini_set('track_errors',$track);
			return $response;
		} else {
			throw new Exception("mediaburstSMS requires PHP5 or PHP4 with cURL");
		}		
	}
}
